sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/NumberUtils.scala

```diff
@@ -164,26 +192,86 @@ object NumberUtils {
-      var resultStr = decimalFormat.format(bigDecimal)
+      val numberDecimalFormat = {
+        val decimalFormat = new DecimalFormat()
+        decimalFormat.setParseBigDecimal(true)
+        try {
+          decimalFormat.applyLocalizedPattern(transformedFormat)
+        } catch {
+          case _: IllegalArgumentException =>
+            throw QueryExecutionErrors.invalidNumberFormatError(numberFormat)
+        }
+        decimalFormat
+      }
+      var resultStr = numberDecimalFormat.format(bigDecimal)
       // Since we trimmed the comma at the beginning or end of number format in function
       // `normalize`, we restore the comma to the result here.
       // For example, if the specified number format is "99,999," or ",999,999", function
       // `normalize` normalize them to "##,###" or "###,###".
       // new DecimalFormat("##,###").parse(12454) and new DecimalFormat("###,###").parse(124546)
       // will return "12,454" and "124,546" respectively. So we add ',' at the end and head of
       // the result, then the final output are "12,454," or ",124,546".
-      if (numberFormat.last == commaSign || numberFormat.last == letterCommaSign) {
-        resultStr = resultStr + commaSign
+      if (numberFormat.last == COMMA_SIGN || numberFormat.last == COMMA_LETTER) {
+        resultStr = resultStr + COMMA_SIGN
       }
-      if (numberFormat.charAt(0) == commaSign || numberFormat.charAt(0) == letterCommaSign) {
-        resultStr = commaSign + resultStr
+      if (numberFormat.charAt(0) == COMMA_SIGN || numberFormat.charAt(0) == COMMA_LETTER) {
+        resultStr = COMMA_SIGN + resultStr
       }
 
       resultStr
     }
   }
 
+  class NumberFormatBuilder(originNumberFormat: String) extends Serializable {
+
+    protected val normalizedNumberFormat = normalize(originNumberFormat)
+
+    private val transformedFormat = transform(normalizedNumberFormat)
+
+    private lazy val numberDecimalFormat = {
+      val decimalFormat = new DecimalFormat()
+      decimalFormat.setParseBigDecimal(true)
+      try {
+        decimalFormat.applyLocalizedPattern(transformedFormat)
+      } catch {
+        case _: IllegalArgumentException =>
+          throw QueryExecutionErrors.invalidNumberFormatError(originNumberFormat)
+      }
+      decimalFormat
+    }
+
+    private val precision = getPrecision(normalizedNumberFormat)
+
+    private val scale = getScale(normalizedNumberFormat)
+
+    private var unChecked = true
+
+    def parsePrecisionAndScale(): (Int, Int) = (precision, scale)
+
+    def check(): TypeCheckResult = {
+      try {
+        NumberUtils.check(normalizedNumberFormat, originNumberFormat)
+      } catch {
+        case e: AnalysisException => return TypeCheckResult.TypeCheckFailure(e.getMessage)
+      }
+      unChecked = false
+      TypeCheckResult.TypeCheckSuccess
+    }
+
+    def parse(input: UTF8String): Decimal = {
+      if (unChecked) {
+        check()
```

Comment:
I'm confused. This is a static type check, why do you do it at runtime when parsing the first input?